File: ADT/tries/648.cpp
class Solution
{
public:
    struct trieNode
    {
        trieNode *children[26];
        bool isEnd;
    };

    trieNode *getNode()
    {
        trieNode *newNode = new trieNode();
        newNode->isEnd = false;

        for (int i = 0; i < 26; i++)
        {
            newNode->children[i] = NULL;
        }
        return newNode;
    }
    trieNode *root;
    void insert(string word)
    {
        trieNode *crawler = root;
        for (int i = 0; i < word.length(); i++)
        {
            int idx = word[i] - 'a';
            if (crawler->children[idx] == NULL)
            {
                crawler->children[idx] = getNode();
            }
            crawler = crawler->children[idx];
        }
        crawler->isEnd = true;
    }

    string search(string word)
    {
        trieNode *crawler = root;
        // O(L) ; l = length
        for (int i = 0; i < word.length(); i++)
        {
            int idx = word[i] - 'a';
            if (crawler->children[idx] == NULL)
            {
                return word;
            }
            crawler = crawler->children[idx];
            if (crawler->isEnd)
            {
                return word.substr(0, i + 1);
            }
        }
        return word;
    }
    string replaceWords(vector<string> &dictionary, string sentence)
    {
        root = getNode();
        // n words
        // l length
        // n * l = total characters of string int the dictionary
        // TC = O(n*l)
        // SC = O(n*l)
        for (string &word : dictionary)
        {
            insert(word);
        }
        stringstream ss(sentence);
        string word;
        string result;
        while (getline(ss, word, ' ')) // O(W * L)
        {
            result += search(word) + ' ';
        }
        result.pop_back();
        return result;
    }
};

File: ADT/tries/648.js
class TrieNode {
    constructor() {
        this.children = new Array(26).fill(null);
        this.isEnd = false;
    }
}

class Trie {
    constructor() {
        this.root = this.getNode();
    }

    getNode() {
        return new TrieNode();
    }

    insert(word) {
        let crawler = this.root;
        for (let char of word) {
            const idx = char.charCodeAt(0) - 'a'.charCodeAt(0);
            if (crawler.children[idx] === null) {
                crawler.children[idx] = this.getNode();
            }
            crawler = crawler.children[idx];
        }
        crawler.isEnd = true;
    }

    search(word) {
        let crawler = this.root;
        for (let i = 0; i < word.length; i++) {
            const char = word[i];
            const idx = char.charCodeAt(0) - 'a'.charCodeAt(0);
            if (crawler.children[idx] === null) {
                return word;
            }
            crawler = crawler.children[idx];
            if (crawler.isEnd) {
                return word.slice(0, i + 1);
            }
        }
        return word;
    }

    replaceWords(dictionary, sentence) {
        for (let word of dictionary) {
            this.insert(word);
        }
        const words = sentence.split(' ');
        const result = [];
        for (let word of words) {
            result.push(this.search(word));
        }
        return result.join(' ');
    }
}